btrack/io/_localization.py

```diff
@@ -16,119 +18,184 @@
+        if key not in nodes:
+            nodes[key] = values
+        else:
+            nodes[key] = np.concatenate([nodes[key], values])
+    return nodes
 
-    # if class id is specified then extract that property first
-    if assign_class_ID:
-        # ensure regionprops can properly read label image
-        labeled = label(segmentation)
 
-        # pull class_ID from segments using pixel intensity
-        _class_ID_centroids = regionprops_table(
-            labeled,
-            intensity_image=segmentation,
-            properties=("max_intensity",),
-        )
+@dataclasses.dataclass
+class SegmentationContainer:
+    """Container for segmentation data."""
+
+    segmentation: Union[Generator, npt.NDArray]
+    intensity_image: Optional[Union[Generator, npt.NDArray]] = None
 
-        # rename class_ID column and remove keyword from properties
-        _class_ID_centroids["class_id"] = _class_ID_centroids.pop(
-            "max_intensity"
+    def __post_init__(self) -> None:
+        self._is_generator = isinstance(self.segmentation, Generator)
+        self._next = (
+            self._next_generator if self._is_generator else self._next_array
         )
 
-        # run regionprops to record other intensity image properties
-        _centroids = regionprops_table(
-            labeled,
-            intensity_image=intensity_image,
-            properties=properties,
+    def _next_generator(self) -> Tuple[npt.NDArray, Optional[npt.NDArray]]:
+        """__next__ method for a generator input."""
+        seg = next(self.segmentation)
+        intens = (
+            next(self.intensity_image)
+            if self.intensity_image is not None
+            else None
+        )
+        return seg, intens
+
+    def _next_array(self) -> Tuple[npt.NDArray, Optional[npt.NDArray]]:
+        """__next__ method for an array-like input."""
+        if self._iter >= len(self):
+            raise StopIteration
+        seg = np.asarray(self.segmentation[self._iter, ...])
+        intens = (
+            np.asarray(self.intensity_image[self._iter, ...])
+            if self.intensity_image is not None
+            else None
         )
+        return seg, intens
+
+    def __iter__(self) -> SegmentationContainer:
+        self._iter = 0
+        return self
+
+    def __next__(self) -> Tuple[int, npt.NDArray, Optional[npt.NDArray]]:
+        seg, intens = self._next()
+        data = (self._iter, seg, intens)
+        self._iter += 1
+        return data
+
+    def __len__(self) -> int:
+        if not self._is_generator:
+            return self.segmentation.shape[0]
+        return 0
+
+
+@dataclasses.dataclass
+class NodeProcessor:
+    """Processor to extract nodes from a segmentation image."""
+
+    properties: Tuple[str]
+    centroid_type: str = "centroid"
+    intensity_image: Optional[npt.NDArray] = None
+    scale: Optional[Tuple[float]] = None
+    assign_class_ID: bool = False  # noqa: N815
```

Comment:
Once `python=3.10` is the min we support, we can use `dataclasses.KW_ONLY` to avoid the `noqa` here

Reply:
Nice! I think this is actually because of the mixed case here

Reply:
Oops, I should have checked what the error is for